### PR TITLE
Close websockets when app is terminated

### LINKED ISSUE #34: Close WebSocket connections on exit
from ipkn/crow#377

The problem with some of the comments there though is that they use line numbers in crow_all.h which means that the smallest change in any header file puts that number way off. Best to use method names or namespaces.

### PR DESCRIPTION
**Pull request by The-EDev**
As mentioned in #34. Crow doesn't properly close websocket connections when terminated, instead closing the socket directly. Which could cause a problem with websocket clients if they're not configured to handle unclean closures.

This *Draft* PR blocks the main thread until all websocket connections have been cleanly closed (which closes #34).

It does so by keeping track of the number of currently open websockets, which is incremented up once the connection object is created, and decremented whenever it is destroyed. The `SIGINT` or `SIGTERM` is sent to both the connection and the server. the server's `do_accept` method is halted and the thread is then blocked until the `websocket_count` reaches `0`, the connections just call the `close()` method, which automatically decrements the count once complete.

There are a few things I'm not comfortable with in the PR:
1. ~The fact that I'm sending the websocket count from the App through the router's `handle_upgrade` method (which makes little sense) and to the websocket connection.~
2. ~Continuing the subject of sending metadata, I've (for now) hardcoded the signals into the `websocket::connection` constructor, which makes `Crow::signal_clear()` and `Crow::signal_add()` methods completely useless, since they don't affect the websocket connections.~
3. There's a problem where deleting a websocket connection and attempting to call `io_service.stop()` on the related IO Context causes a `SEGFAULT`. (More details below)
4. I seem to have added a bunch of spaces and tabs where they shouldn't be.

The fourth problem is no big deal. The other 3 however I do need help in.

~To deal with numbers 1 & 2 my Idea is to create a new class named either `crow::connection_context` or `crow::metadata` which would contain the websocket count, signals, and potentially other data that the App needs to send to connections, adaptors, etc..
I didn't work on it because I need other opinions since this is an architectural change.~

The third issue is quite a bit more complicated. Here's the text I wrote (on [Gitter](https://gitter.im/crowfork/community)) while looking into it:
> I got websockets to close when the app exits, problem is that I seem to have uncovered a bigger bug doing this. It seems that stopping io_contexts (io_services) references connection objects, and websockets get deleted once they're closed, therefore causing a segfault. I need to understand how io_contexts work and how to fix this issue before I even think of committing the rest of the code

> One theory I had was that the adaptor(`ASIO::TCP::SOCKET`) was somehow being referenced by the `io_service` as it was getting stopped, and since it was owned by the `websocket::Connection`, deleting it would cause the `io_service` to attempt to dereference what is now a `null` pointer.
My hypothesis was proven correct when I separated the adaptor and only deleted it, leaving the connection intact. And got the same `SEGFAULT`.
However I was proven wrong when I did the opposite, deleting only the connection and leaving the adaptor intact. I still got `SEGFAULT`
It seems that for whatever reason both need to be in memory for things to close properly, and I have no idea why.

I've already worked around the issue using the `should_close_` boolean. Which prevents the websocket object from being deleted after being closed. I particularly don't like this because A. it ignores the base issue that something that doesn't exist is being referenced. and B. It potentially causes a memory leak since there's no actual reference to the websocket connection:
Here's the code responsible for creating websocket connections (notice how it is created in the heap and left to deal with itself)
https://github.com/CrowCpp/Crow/blob/b5137c52a1323fb4793bbebf4a48b528c939bc98/include/crow/routing.h#L403-L406

### CODE REVIEW

**Comment by The-EDev** [comment]
CI build killed because more work needs to be done anyway.

**Comment by The-EDev** [comment]
I seriously should NOT test this here..

**Comment by The-EDev** [comment]
> However I was proven wrong when I did the opposite, deleting only the connection and leaving the adaptor intact. I still got SEGFAULT

This makes sense because the adaptor is moved to be a member of the websocket connection, thus deleting the connection would also delete the adaptor, which is most likely the reason behind the segfault.

**Comment by The-EDev** [comment]
Adding `adaptor_.shutdown_readwrite();` seems to resolve the segfault issue..

**Comment by The-EDev** [comment]
@mrozigor @luca-schlecker I believe this PR is ready to be merged, Could you please take a look at it? Thanks.

File: include/crow/http_server.h [hunk 1]
```diff
@@ -161,9 +161,21 @@ namespace crow
 
         void stop()
         {
-            io_service_.stop();
+            should_close_ = false;                                                 //Prevent the acceptor from taking new connections
```
Comment (mrozigor) [inline include/crow/http_server.h:164]: Is it used somewhere?

Comment (The-EDev) [reply include/crow/http_server.h:164]: I must have missed adding functionality to this when I was first working on the feature. The comment seems to indicate that setting this to `true` would disable `do_accept()`. I'm pushing a commit to get that working now.

File: include/crow/http_server.h [hunk 2]
```diff
@@ -207,33 +207,36 @@ namespace crow
...
-                  }
-                  else
-                  {
-                      task_queue_length_pool_[service_idx]--;
-                      CROW_LOG_DEBUG << &is << " {" << service_idx << "} queue length: " << task_queue_length_pool_[service_idx];
-                      delete p;
-                  }
-                  do_accept();
-              });
+            if (!should_close_)
```
Comment (mrozigor) [inline include/crow/http_server.h:210]: Maybe change name to `working` and invert states? Then it should be easier to understand (no need to invert) ;)

Comment (The-EDev) [reply include/crow/http_server.h:210]: what about something like `shutting_down_`? Since the boolean is only true while Crow is closing

Comment (mrozigor) [reply include/crow/http_server.h:210]: Hmm, it need to be inverted, but is IMHO more readable than the current version ;)

Comment (The-EDev) [reply include/crow/http_server.h:210]: I believe I already inverted the states in the past commit, I'll push the name change now :)

**Review by mrozigor — APPROVED** [review]

**Pull request merged**